### PR TITLE
fix(networkfirewall): fixed import for networkfirewall resources

### LINKED ISSUE #657: `aws_networkfirewall_firewall_policy` and `aws_networkfirewall_rule_group` import not possible
### What happened?
for the following resources no import is possible - this is a blocker for migration to official provider-aws

`aws_networkfirewall_firewall_policy`
`aws_networkfirewall_rule_group`

the import leads to the following error:

`    Message:               apply failed: creating NetworkFirewall Rule Group (rulegroup): InvalidRequestException: A resource with the specified name already exists: `

`    Message:               apply failed: creating NetworkFirewall Firewall Policy (firewallpolicy): InvalidRequestException: A resource with the specified name already exists: `

we have also the following issue:
https://github.com/upbound/provider-aws/blob/main/config/externalname.go#L2239 - this is wrong because we have a difference between stateful-rulegroup and stateless-rule group in the ARN - think its combined to the general import problem 

### How can we reproduce it?

`kubectl create -f firewall.yaml `

check that all resources created:
```
kubectl get managed
NAME                                                 READY   SYNCED   EXTERNAL-NAME   AGE
rulegroup.networkfirewall.aws.upbound.io/rulegroup   True    True     rulegroup       9m16s

NAME                                                           READY   SYNCED   EXTERNAL-NAME    AGE
firewallpolicy.networkfirewall.aws.upbound.io/firewallpolicy   True    True     firewallpolicy   9m16s

NAME                                                   READY   SYNCED   EXTERNAL-NAME                                                           AGE
firewall.networkfirewall.aws.upbound.io/firewallname   True    True     arn:aws:network-firewall:us-west-1:123456789101:firewall/firewallname   9m16s

NAME                                       READY   SYNCED   EXTERNAL-NAME              AGE
subnet.ec2.aws.upbound.io/sample-subnet1   True    True     subnet-07c67cb02a111a809   9m16s

NAME                                READY   SYNCED   EXTERNAL-NAME           AGE
vpc.ec2.aws.upbound.io/sample-vpc   True    True     vpc-0707128cc333d611e   9m16s
```

export the following resources:
```
kubectl get rulegroup.networkfirewall.aws.upbound.io/rulegroup -o yaml > rulegroup-orphan.yaml 
kubectl get firewallpolicy.networkfirewall.aws.upbound.io/firewallpolicy -o yaml > firewallpolicy-orphan.yaml
```

set both resources to `deletionPolicy: Orphan`

```
kubectl apply -f rulegroup-orphan.yaml 
kubectl apply -f firewallpolicy-orphan.yaml
```

remove both resources

```
kubectl delete -f rulegroup-orphan.yaml 
kubectl delete -f firewallpolicy-orphan.yaml
```

create both resources:

```
kubectl apply -f rulegroup-orphan.yaml 
kubectl apply -f firewallpolicy-orphan.yaml
```

check the status 
```
kubectl get managed
NAME                                                           READY   SYNCED   EXTERNAL-NAME    AGE
firewallpolicy.networkfirewall.aws.upbound.io/firewallpolicy   False   True     firewallpolicy   21m

NAME                                                 READY   SYNCED   EXTERNAL-NAME   AGE
rulegroup.networkfirewall.aws.upbound.io/rulegroup   False   True     rulegroup       21m
```

```
    Message:               apply failed: creating NetworkFirewall Rule Group (rulegroup): InvalidRequestException: A resource with the specified name already exists: 

    Message:               apply failed: creating NetworkFirewall Firewall Policy (firewallpolicy): InvalidRequestException: A resource with the specified name already exists: 
```

===

### What environment did it happen in?

* Crossplane Version: 1.11.3
* Provider Version: v0.31.0
* Kubernetes Version:  
* Kubernetes Distribution:  EKS 1.23

### LINKED ISSUE #658: `aws_networkfirewall_firewall_policy` missing ref/selector for RuleGroup
### What happened?

`aws_networkfirewall_firewall_policy ` resources is missing ref/selector for RuleGroup  in `firewall_policy.stateless_rule_group_reference.resource_arn` and `firewall_policy.stateful_rule_group_reference.resource_arn`

```
error: error validating "default-policy.yaml": error validating data: [ValidationError(FirewallPolicy.spec.forProvider.firewallPolicy[0].statefulRuleGroupReference[0]): unknown field "resourceArnRef" in io.upbound.aws.networkfirewall.v1beta1.FirewallPolicy.spec.forProvider.firewallPolicy.statefulRuleGroupReference, ValidationError(FirewallPolicy.spec.forProvider.firewallPolicy[0].statefulRuleGroupReference[1]): unknown field "resourceArnRef" in io.upbound.aws.networkfirewall.v1beta1.FirewallPolicy.spec.forProvider.firewallPolicy.statefulRuleGroupReference]; if you choose to ignore these errors, turn validation off with --validate=false
```

### How can we reproduce it?
create the following manifest:

```yaml
apiVersion: networkfirewall.aws.upbound.io/v1beta1
kind: FirewallPolicy
metadata:
  name: default-policy
spec:
  deletionPolicy: Orphan
  forProvider:
    region: eu-central-1
    name: default-policy
    firewallPolicy:
      - statelessDefaultActions:
          - "aws:forward_to_sfe"
        statelessFragmentDefaultActions:
          - "aws:pass"
        statelessRuleGroupReference:
          - resourceArnRef:
              name: forward-all
            priority: 2
        statefulRuleGroupReference:
          - resourceArnRef:
              name: allow-domainlist
          - resourceArnRef:
              name: deny
```

### What environment did it happen in?

* Crossplane Version: 1.11.3
* Provider Version: 0.31.0
* Kubernetes Version:  
* Kubernetes Distribution:  EKS 1.23

### PR DESCRIPTION
**Pull request by haarchri**
### Description of your changes
1) fix import for the following networkfirewall resources:
`aws_networkfirewall_firewall_policy`
`aws_networkfirewall_rule_group`

2)implement missing ref/selector for ruleGroup in `aws_networkfirewall_firewall_policy`
3)add more complex example for `aws_networkfirewall_firewall_policy` and `aws_networkfirewall_rule_group`

Fixes #658 
Fixes #657

I have:

- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested
```
kubectl apply -f firewallpolicy-complex.yaml
firewallpolicy.networkfirewall.aws.upbound.io/example-default-policy created
rulegroup.networkfirewall.aws.upbound.io/example-forward-all created
rulegroup.networkfirewall.aws.upbound.io/example-allow-domainlist created
rulegroup.networkfirewall.aws.upbound.io/example-deny created
```

```
NAME                                                                   READY   SYNCED   EXTERNAL-NAME            AGE
firewallpolicy.networkfirewall.aws.upbound.io/example-default-policy   True    True     example-default-policy   3m17s

NAME                                                                READY   SYNCED   EXTERNAL-NAME              AGE
rulegroup.networkfirewall.aws.upbound.io/example-allow-domainlist   True    True     example-allow-domainlist   3m17s
rulegroup.networkfirewall.aws.upbound.io/example-deny               True    True     example-deny               3m17s
rulegroup.networkfirewall.aws.upbound.io/example-forward-all        True    True     example-forward-all        3m17s
```

updated `deletionPolicy: Orphan`

```
kubectl delete -f firewallpolicy-complex.yaml 
firewallpolicy.networkfirewall.aws.upbound.io "example-default-policy" deleted
rulegroup.networkfirewall.aws.upbound.io "example-forward-all" deleted
rulegroup.networkfirewall.aws.upbound.io "example-allow-domainlist" deleted
rulegroup.networkfirewall.aws.upbound.io "example-deny" deleted
```

```
kubectl apply -f firewallpolicy-complex.yaml
firewallpolicy.networkfirewall.aws.upbound.io/example-default-policy created
rulegroup.networkfirewall.aws.upbound.io/example-forward-all created
rulegroup.networkfirewall.aws.upbound.io/example-allow-domainlist created
rulegroup.networkfirewall.aws.upbound.io/example-deny created
```

```
NAME                                                                   READY   SYNCED   EXTERNAL-NAME            AGE
firewallpolicy.networkfirewall.aws.upbound.io/example-default-policy   True    True     example-default-policy   1m12s

NAME                                                                READY   SYNCED   EXTERNAL-NAME              AGE
rulegroup.networkfirewall.aws.upbound.io/example-allow-domainlist   True    True     example-allow-domainlist   1m12s
rulegroup.networkfirewall.aws.upbound.io/example-deny               True    True     example-deny               1m12s
rulegroup.networkfirewall.aws.upbound.io/example-forward-all        True    True     example-forward-all        1m12s
```

### CODE REVIEW

**Comment by haarchri** [comment]
cannot add breaking change to this PR :/

**Comment by haarchri** [comment]
/test-examples="examples/networkfirewall/firewall.yaml"

**Comment by haarchri** [comment]
/test-examples="examples/networkfirewall/firewallpolicy-complex.yaml"

File: config/externalname.go [hunk 1]
```diff
@@ -2233,10 +2233,10 @@ var ExternalNameConfigs = map[string]config.ExternalName{
 	//
 	// Network Firewall Policies can be imported using their ARN
 	// Example: arn:aws:network-firewall:us-west-1:123456789012:firewall-policy/example
-	"aws_networkfirewall_firewall_policy": config.TemplatedStringAsIdentifier("name", "arn:aws:network-firewall:{{ .parameters.region }}:{{ .setup.client_metadata.account_id }}:firewall-policy/{{ .external_name }}"),
+	"aws_networkfirewall_firewall_policy": config.IdentifierFromProvider,
 	// Network Firewall Rule Groups can be imported using their ARN
 	// Example: arn:aws:network-firewall:us-west-1:123456789012:stateful-rulegroup/example
-	"aws_networkfirewall_rule_group": config.TemplatedStringAsIdentifier("", "arn:aws:network-firewall:{{ .parameters.region }}:{{ .setup.client_metadata.account_id }}:stateful-rulegroup/{{ .external_name }}"),
+	"aws_networkfirewall_rule_group": config.IdentifierFromProvider,
```
Comment (ytsarev) [inline config/externalname.go:2239]: Can we document in the comment above that the resource id deviates between `stateful-group` and `stateless-group` , that's why we need to switch to `config.IdentifierFromProvider`

File: config/externalname.go [hunk 2]
```diff
@@ -2233,10 +2233,10 @@ var ExternalNameConfigs = map[string]config.ExternalName{
 	//
 	// Network Firewall Policies can be imported using their ARN
 	// Example: arn:aws:network-firewall:us-west-1:123456789012:firewall-policy/example
-	"aws_networkfirewall_firewall_policy": config.TemplatedStringAsIdentifier("name", "arn:aws:network-firewall:{{ .parameters.region }}:{{ .setup.client_metadata.account_id }}:firewall-policy/{{ .external_name }}"),
+	"aws_networkfirewall_firewall_policy": config.IdentifierFromProvider,
```
Comment (ytsarev) [inline config/externalname.go:2236]: Is a switch to `config.IdentifierFromProvider` really required here? I understand the deviation mitigation in the `aws_networkfirewall_rule_group` below, but here it [looks](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/networkfirewall_firewall_policy#import) static . Does it deviate too or there is another reason that is not obvious? . I any case, please document the decision in the comment above

**Review by ytsarev — COMMENTED** [review]
Thanks a lot for the thorough testing. Left couple of comments regarding external name configuration

File: config/externalname.go (hunk 2 above)
Comment (haarchri) [reply config/externalname.go:2236]: we have the following issue during import - is `.setup.....` not working during import ?

`   Message:               apply failed: creating NetworkFirewall Firewall Policy (firewallpolicy): InvalidRequestException: A resource with the specified name already exists:`

**Comment by haarchri** [comment]
@ulucinar  any idea for the issue for import ? the only thing which helped was to switch the identifier - great to hear your feedback ;)

File: config/externalname.go (hunk 2 above)
Comment (sergenyalcin) [reply config/externalname.go:2236]: I am not sure that the root cause is [this issue](https://github.com/upbound/upjet/issues/32) but sometimes we are observing some interesting issues about the custom external name configurations. This can be also related to these type of problems. And when we switch the custom configurations to `IdentifierFromProvider`, they are fixing.

@ytsarev I think I understand your point. Actually, you say that the ID that we construct is the same as the ID that the provider sends us (IdentifierFromProvider case). I think you are right about this. However, I think that the reason for these types of errors, which cannot be understood well from time to time, may be that the import and apply paths on the terraform side differ for some resources.

In short, we observed similar problems for different resources. We may consider investigate more the root cause of this type of problems.

Maybe @ulucinar also want to say his insights on this.

Comment (ytsarev) [reply config/externalname.go:2236]:
```
he import and apply paths on the terraform side differ for some resources.
```
ah, that might be the case and key for understanding the situation, thank you!

Comment (ulucinar) [inline config/externalname.go:2236]: When we introduce the `spec.forProvider.name` back to the firewall policy resource, we will also need to add it to the manually maintained manifests: 
https://github.com/upbound/provider-aws/blob/1faeb9c0b215155b2c62cf13fc1200dd4dabb94d/examples/networkfirewall/firewallpolicy.yaml#L10

File: examples/networkfirewall/firewallpolicy-complex.yaml [hunk 3]
```diff
@@ -0,0 +1,124 @@
+apiVersion: networkfirewall.aws.upbound.io/v1beta1
+kind: FirewallPolicy
+metadata:
+  annotations:
+    meta.upbound.io/example-id: networkfirewall/v1beta1/firewallpolicy-complex
```
Comment (ulucinar) [inline examples/networkfirewall/firewallpolicy-complex.yaml:5]: The example ID specifies a GVK and is important so that we can correctly show the example in the marketplace:
```suggestion
    meta.upbound.io/example-id: networkfirewall/v1beta1/firewallpolicy
```

File: examples/networkfirewall/firewallpolicy-complex.yaml [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    name: example-default-policy
+    region: us-west-1
+
+---
+
+apiVersion: networkfirewall.aws.upbound.io/v1beta1
+kind: RuleGroup
+metadata:
+  annotations:
+    meta.upbound.io/example-id: networkfirewall/v1beta1/firewallpolicy-complex
```
Comment (ulucinar) [inline examples/networkfirewall/firewallpolicy-complex.yaml:37]:
```suggestion
    meta.upbound.io/example-id: networkfirewall/v1beta1/firewallpolicy
```

File: examples/networkfirewall/firewallpolicy-complex.yaml [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+                            - addressDefinition: "0.0.0.0/0"
+    type: STATELESS
+
+---
+
+apiVersion: networkfirewall.aws.upbound.io/v1beta1
+kind: RuleGroup
+metadata:
+  annotations:
+    meta.upbound.io/example-id: networkfirewall/v1beta1/firewallpolicy-complex
```
Comment (ulucinar) [inline examples/networkfirewall/firewallpolicy-complex.yaml:67]:
```suggestion
    meta.upbound.io/example-id: networkfirewall/v1beta1/firewallpolicy
```

File: examples/networkfirewall/firewallpolicy-complex.yaml [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+              generatedRulesType: "ALLOWLIST"
+    type: STATEFUL
+
+---
+
+apiVersion: networkfirewall.aws.upbound.io/v1beta1
+kind: RuleGroup
+metadata:
+  annotations:
+    meta.upbound.io/example-id: networkfirewall/v1beta1/firewallpolicy-complex
```
Comment (ulucinar) [inline examples/networkfirewall/firewallpolicy-complex.yaml:99]:
```suggestion
    meta.upbound.io/example-id: networkfirewall/v1beta1/firewallpolicy
```

**Review by ulucinar — COMMENTED** [review]
Hi @haarchri,
Thank you for addressing these issues. I'm posting some review comments to discuss.

I will also investigate why the old configuration is broken and leave a comment in the discussion.

File: config/externalname.go (hunk 2 above)
Comment (ulucinar) [reply config/externalname.go:2236]: Hi folks,
Sorry for the late answer. Just left a comment [here](https://github.com/upbound/provider-aws/issues/657#issuecomment-1507259764) explaining the root causes of the issues we are observing while importing these resources.

I will propose some different fixes than switching to `config.IdentifierFromProvider` because:
- It will introduce breaking changes for both resources
- We will lose `spec.forProvider` API normalization for the `networkfirewall.FirewallPolicy` resource if we switch to `config.IdentifierFromProvider`, i.e., we will introduce the `spec.forProvider.name` field as a naming field for the resource, which's a violation of the XRM.
- We will lose the external-name normalization if we switch to the `config.IdentifierFromProvider` for `aws_networkfirewall_rule_group`: The external-name is right now just the name of the rule group but it will become a full ARN if we make the switch. However, it looks like we need a new feature in upjet's `config.TemplatedStringAsIdentifier`, which I will propose in the review comment suggesting a change. 

Let's discuss, and thank you.

Comment (ulucinar) [inline config/externalname.go:2236]: I've successfully verified the following with the import scenario:
```suggestion
	"aws_networkfirewall_firewall_policy": config.TemplatedStringAsIdentifier("name", "arn:aws:network-firewall:{{ .setup.configuration.region }}:{{ .setup.client_metadata.account_id }}:firewall-policy/{{ .external_name }}"),
```

Why I'm suggesting this is because we keep the API normalized as discussed in some depth [here](https://github.com/upbound/provider-aws/pull/661#discussion_r1165788746).

File: config/externalname.go (hunk 1 above)
Comment (ulucinar) [inline config/externalname.go:2239]: As discussed [here](https://github.com/upbound/provider-aws/issues/657#issuecomment-1507259764), we have two problems with the old configuration. We can replace the `.parameters.region` in the template expression with `{{ .setup.configuration.region }}` as suggested above but this will not be sufficient because we also need to take care of the `stateful-rulegroup` ARN component. 

However, if we extend the templating functions available for `config.TemplatedStringAsIdentifier`, we will be able to do the following:
```suggestion
	"aws_networkfirewall_rule_group": config.TemplatedStringAsIdentifier("", "arn:aws:network-firewall:{{ .setup.configuration.region }}:{{ .setup.client_metadata.account_id }}:{{ .parameters.type | ToLower }}-rulegroup/{{ .external_name }}"),
```

This will also ensure external-name normalization and backward-compatibility as discussed [here](https://github.com/upbound/provider-aws/pull/661#discussion_r1165788746). 

If we agree on this, I can open a small PR in upjet adding the `ToLower` templating function to make this possible. Also verified this configuration with imports on my local setup.

**Review by ulucinar — COMMENTED** [review]
Hi @haarchri,
After the [root cause analysis](https://github.com/upbound/provider-aws/issues/657#issuecomment-1507259764), I'm leaving some more comments for us to discuss. If we do not switch to `config.IdentifierFromProvider`, then we will also need to change the extractors in the newly introduced referencer configuration. Currently, those new referencer configurations are relying on the rule group being configured with `config.IdentifierFromProvider`. We may use the [`common.PathARNExtractor`](https://github.com/upbound/provider-aws/blob/1faeb9c0b215155b2c62cf13fc1200dd4dabb94d/config/common/common.go#L20) to extract the ARN of a rule group resource.

File: config/externalname.go (hunk 1 above)
Comment (haarchri) [reply config/externalname.go:2239]: > As discussed [here](https://github.com/upbound/provider-aws/issues/657#issuecomment-1507259764), we have two problems with the old configuration. We can replace the `.parameters.region` in the template expression with `{{ .setup.configuration.region }}` as suggested above but this will not be sufficient because we also need to take care of the `stateful-rulegroup` ARN component. 
> 
> However, if we extend the templating functions available for `config.TemplatedStringAsIdentifier`, we will be able to do the following:
> ```suggestion
> 	"aws_networkfirewall_rule_group": config.TemplatedStringAsIdentifier("", "arn:aws:network-firewall:{{ .setup.configuration.region }}:{{ .setup.client_metadata.account_id }}:{{ .parameters.type | ToLower }}-rulegroup/{{ .external_name }}"),
> ```
> 
> This will also ensure external-name normalization and backward-compatibility as discussed [here](https://github.com/upbound/provider-aws/pull/661#discussion_r1165788746). 
> 
> If we agree on this, I can open a small PR in upjet adding the `ToLower` templating function to make this possible. Also verified this configuration with imports on my local setup.

yes open the PR in upjet and provide the new version i can update this PR

Comment (ulucinar) [reply config/externalname.go:2239]: Once we consume https://github.com/upbound/upjet/pull/190, we can proceed with the suggested configuration.

Comment (ulucinar) [reply config/externalname.go:2239]: Hi @haarchri,
https://github.com/upbound/upjet/pull/190 has been merged. We can bump the upjet dependency in this PR and proceed with the configuration changes if you'd like. The template functions are available in commit `1268a48eba95ef194823f4bcc118e14a5d867a94` and later. Thank you very much.

Comment (haarchri) [reply config/externalname.go:2239]: Start working on it

**Comment by haarchri** [comment]
@ulucinar  we can remove the breaking-change label then

**Comment by haarchri** [comment]
/test-examples="examples/networkfirewall/firewallpolicy-complex.yaml"

**Review by ulucinar — APPROVED** [review]
Thank you @haarchri, lgtm.
